hazelcast/src/main/java/com/hazelcast/query/PartitionPredicate.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.query;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.DataSerializable;
+
+import java.io.IOException;
+import java.util.Map;
+
+import static com.hazelcast.util.Preconditions.checkNotNull;
+
+/**
+ * A {@link Predicate} that restricts the execution of a Predicate to a single Partition. This can help to speed up query
+ * execution since only 1 instead of all partitions needs to be queried.
+ *
+ * This predicate can only be used as an outer predicate.
+ *
+ * @param <K>
+ * @param <V>
+ */
+public class PartitionPredicate<K, V> implements Predicate<K, V>, DataSerializable {
+
+    private Object partitionKey;
+    private Predicate<K, V> target;
+
+    private PartitionPredicate() {
+    }
+
+    /**
+     * Creates a new PartitionPredicate
+     *
+     * @param partitionKey the partition key
+     * @param target       the target Predicate
+     * @throws IllegalArgumentException if partitionId smaller than 0
+     * @throws NullPointerException     if target Predicate is null.
+     */
+    public PartitionPredicate(Object partitionKey, Predicate<K, V> target) {
+        this.partitionKey = checkNotNull(partitionKey, "partitionKey can't be null");
+        this.target = checkNotNull(target, "target predicate can't be null");
+    }
+
+    /**
+     * The partition key that determines the partition the target Predicate is going to execute on.
+     *
+     * @return the partition id.
+     */
+    public Object getPartitionKey() {
+        return partitionKey;
+    }
+
+    /**
+     * Returns the target Predicate.
+     *
+     * @return the target Predicate.
+     */
+    public Predicate<K, V> getTarget() {
+        return target;
+    }
+
+    @Override
+    public boolean apply(Map.Entry<K, V> mapEntry) {
+        throw new UnsupportedOperationException();
```

Comment:
The PartitionPredicate is meant as a control mechanism for the QueryEngine to schedule the Query. This PartitionPredicate is then stripped away from the actual predicate and that gets executed.

If this PartitionPredicate does allow for execution, it would be possible to do
```
not(partitionpredicate(1,true))
```
But this would not be valid, since it would not be executing on partition one. So it is best to fail the PartitionPredicate if it does get executed. This would follow the fail fast principle.